---
title: "Springboard Foundations of Data Science"
author: "Rocio Dominguez Vidana, PhD; and Matt Fornito, PhD (mentor)"
date: "August 30, 2016"
output: html_document
---

```{r setup and load data, include=FALSE}

knitr::opts_chunk$set(echo = TRUE)


```

# Development of Gene Expression Profile signature to predict survival in Sarcoma patients (Capstone Project)

Molecular diagnostics is the area at the center of a healthcare revolution and under intense development. It is possible by the increasing use  of powerful profiling tools that leverage our understanding of genetic variability within individuals in healthy and diseased states, and it can inform treatment and prognosis for patients.

The client company has extensive experience developing prognostic tests with “gene expression signatures” in a subset of rare neoplastic diseases for which there are not a lot of research resources allocated to. Using a combination of publically available data curated with further validation. They have developed products which can be used to predict metastatic potential, and thus inform aggressivity of treatment. The company's objective is to achieve that these become “standard of care” by improving the patient's diagnostic; and thus, be rutinarily reimbursed by insurance companies.

The client company has decided to focus on developing a product for sarcomas, a rare type of cancer in humans that arises from transformed cells of mesenchymal origin (bone, cartilage, fat, muscle, vascular, hematopoietic tissues, etc). Within sarcomas there is extreme genetic variability, 30% of them have distinct genetic features that can be considered the cause of those particular tumors. However, for the other 70% percent, there is really no traceable genetic modification, thus suggesting that the driving factor is certain alteration in [genetic expression](https://en.wikipedia.org/wiki/Gene_expression); a hypothesis for which we would need to identify the responsible mRNAs.

Currently, there are two publically available datasets:

* RNASeq TCGA-SARC dataset from the [Cancer Genome Browser](https://genome-cancer.ucsc.edu/proj/site/hgHeatmap/). The dataset TCGA_SARC_exp_HiSeqV2-2015-02-24.tgz is in log2(normalized_count+1) units, and is henceforth referred as the "tcga dataset".

* [RNA-seq performed on sarcomas to identify various alterations](http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE71119)  This dataset belongs to the following [paper](http://www.ncbi.nlm.nih.gov/pubmed/26916546), and is hereafter referred to as the "cinsarc dataset".

Load the two available RNA expression datasets, and the function script.

```{r initial data exploration}

load("capstone.RData")

source("capstone.code.R")

dim(expression.GSE71119)

dim(expression.TCGA.SARC)

plot.cinsarc <- ggplot.histogram(expression.GSE71119, title = "RNASeq Expression CINSARC data")

plot.tcga <- ggplot.histogram(expression.TCGA.SARC, title = "RNASeq Expression TCGA data")

grid.arrange(plot.cinsarc, plot.tcga, ncol=2)

```

There are different subsets of genes available in each dataset, in order to determine the best predictors accross both datasets, we will keep only the common genes. 

```{r identify common genes}

common.expression.genes <- intersect(colnames(expression.TCGA.SARC),colnames(expression.GSE71119))

```

A transformation to a normal distribution by removing the average effect of both all the genes, and all the samples.

```{r normalize datasets}

normalized.tcga <- normalize.expression.data(expression.TCGA.SARC[,common.expression.genes])

normalized.cinsarc <- normalize.expression.data(expression.GSE71119[,common.expression.genes])

plot.cinsarc <- ggplot.histogram(normalized.tcga, title = "Normalized RNASeq CINSARC")

plot.tcga <- ggplot.histogram(normalized.cinsarc, title = "Normalized RNASeq TCGA")

grid.arrange(plot.cinsarc, plot.tcga, ncol=2)

```

Then we will take the top 2/3 of the genes with the highest standard deviation, and we remove the less expressed genes (only the bottom 10% that has a gene mean expression z score lesser than  .5.

```{r select common normalized genes}

selected.tcga <- select.variable.genes (normalized.tcga)

selected.cinsarc <- select.variable.genes (normalized.cinsarc)

common.variable.genes<-intersect(selected.tcga$genes,selected.cinsarc$genes)

```

Then we remove all the genes with correlation higher than .5

```{r choose workable genes, reduce dimensionality}

valid.genes.tcga <- remove.correlated.genes.1(list(genes.rank = selected.tcga$genes.rank, expression.data = normalized.tcga[,common.variable.genes]))

valid.genes.cinsarc <- remove.correlated.genes.1(list(genes.rank = selected.cinsarc$genes.rank, expression.data = normalized.cinsarc[,common.variable.genes]))

common.valid.genes<-intersect(valid.genes.tcga,valid.genes.cinsarc)

for.modeling.expression.cinsarc <- normalized.cinsarc[,common.valid.genes]

for.modeling.expression.tcga <- normalized.tcga[,common.valid.genes]

plot.cinsarc <- ggplot.histogram(for.modeling.expression.cinsarc, title = "Selected RNASeq CINSARC")

plot.tcga <- ggplot.histogram(for.modeling.expression.cinsarc, title = "Selected RNASeq TCGA")

grid.arrange(plot.cinsarc, plot.tcga, ncol=2)


```

```{r delete temporary variables 1, include=FALSE}

rm(common.expression.genes,common.valid.genes,common.variable.genes,selected.cinsarc,selected.tcga,valid.genes.tcga,valid.genes.cinsarc,plot.cinsarc,plot.tcga)

gc()

```

Once we have generated equivalent and comparable matrices for both RNASeq expression datasets, we are shifting our focus to the clinical data. Unfortunately there is not a lot of common information between both patient cohorts, so our focus will be on typing in the same way the output of each individual patient, and homologate as much as possible the pathological information available. This part has been mostly pedestrian and it is not possible to automate it extensively as the management of the expression data.

```{r data janitoring}

clinical.cinsarc <- clinical.GSE71119[,c(8:10)]
colnames(clinical.cinsarc) <- c("subtype","event","time")
clinical.cinsarc$subtype <- as.character(clinical.cinsarc$subtype)
clinical.cinsarc$event <- as.character(clinical.cinsarc$event)
clinical.cinsarc$time <- as.numeric(as.character(clinical.cinsarc$time))*30

clinical.tcga <- clinical.TCGA.SARC[,c(2,9,32)]
colnames(clinical.tcga) <- c("event","time","subtype")
clinical.tcga$event <- as.factor(clinical.tcga$event)
clinical.tcga$time <- as.numeric(clinical.tcga$time)
clinical.tcga$subtype <- as.character(clinical.tcga$subtype)

clinical.cinsarc$subtype[which(clinical.cinsarc$subtype=="Dedifferentiated liposarcoma")]<-rep("1",length(which(clinical.cinsarc$subtype=="Dedifferentiated liposarcoma")))
clinical.cinsarc$subtype[which(clinical.cinsarc$subtype=="Leiomyosarcoma")]<-rep("2",length(which(clinical.cinsarc$subtype=="Leiomyosarcoma")))
clinical.cinsarc$subtype[which(clinical.cinsarc$subtype=="GIST")]<-rep("2",length(which(clinical.cinsarc$subtype=="GIST")))
clinical.cinsarc$subtype[which(clinical.cinsarc$subtype=="Myxofibrosarcoma")]<-rep("3",length(which(clinical.cinsarc$subtype=="Myxofibrosarcoma")))
clinical.cinsarc$subtype[which(clinical.cinsarc$subtype=="Undifferentiated sarcoma")]<-rep("4",length(which(clinical.cinsarc$subtype=="Undifferentiated sarcoma")))
clinical.cinsarc$subtype[which(clinical.cinsarc$subtype=="Other")]<-rep("6",length(which(clinical.cinsarc$subtype=="Other")))
clinical.cinsarc$subtype <- as.integer(clinical.cinsarc$subtype)

clinical.cinsarc$event[which(clinical.cinsarc$event=="Yes")]<-rep("1",length(which(clinical.cinsarc$event=="Yes")))
clinical.cinsarc$event[which(clinical.cinsarc$event=="No")]<-rep("0",length(which(clinical.cinsarc$event=="No")))
clinical.cinsarc$event <- as.factor(clinical.cinsarc$event)

clinical.tcga$subtype[which(clinical.tcga$subtype=="Dedifferentiated liposarcoma")]<-rep("1",length(which(clinical.tcga$subtype=="Dedifferentiated liposarcoma")))
clinical.tcga$subtype[which(clinical.tcga$subtype=="Leiomyosarcoma (LMS)")]<-rep("2",length(which(clinical.tcga$subtype=="Leiomyosarcoma (LMS)")))
clinical.tcga$subtype[which(clinical.tcga$subtype=="Myxofibrosarcoma")]<-rep("3",length(which(clinical.tcga$subtype=="Myxofibrosarcoma")))
clinical.tcga$subtype[which(clinical.tcga$subtype=="Pleomorphic MFH/ Undifferentiated pleomorphic sarcoma")]<-rep("4",length(which(clinical.tcga$subtype=="Pleomorphic MFH/ Undifferentiated pleomorphic sarcoma")))
clinical.tcga$subtype[which(clinical.tcga$subtype=="Giant cell MFH / Undifferentiated pleomorphic sarcoma with giant cells")]<-rep("4",length(which(clinical.tcga$subtype=="Giant cell MFH / Undifferentiated pleomorphic sarcoma with giant cells")))
clinical.tcga$subtype[which(clinical.tcga$subtype=="Undifferentiated Pleomorphic Sarcoma (UPS)")]<-rep("4",length(which(clinical.tcga$subtype=="Undifferentiated Pleomorphic Sarcoma (UPS)")))
clinical.tcga$subtype[which(clinical.tcga$subtype=="Sarcoma; synovial; poorly differentiated")]<-rep("5",length(which(clinical.tcga$subtype=="Sarcoma; synovial; poorly differentiated")))
clinical.tcga$subtype[which(clinical.tcga$subtype=="Synovial Sarcoma - Biphasic")]<-rep("5",length(which(clinical.tcga$subtype=="Synovial Sarcoma - Biphasic")))
clinical.tcga$subtype[which(clinical.tcga$subtype=="Synovial Sarcoma - Monophasic")]<-rep("5",length(which(clinical.tcga$subtype=="Synovial Sarcoma - Monophasic")))
clinical.tcga$subtype[which(clinical.tcga$subtype=="Desmoid Tumor")]<-rep("6",length(which(clinical.tcga$subtype=="Desmoid Tumor")))
clinical.tcga$subtype[which(clinical.tcga$subtype=="Malignant Peripheral Nerve Sheath Tumors (MPNST)")]<-rep("6",length(which(clinical.tcga$subtype=="Malignant Peripheral Nerve Sheath Tumors (MPNST)")))
clinical.tcga$subtype <- as.integer(clinical.tcga$subtype)

```


```{r delete temporary variables 2, include=FALSE}

rm(clinical.GSE71119,clinical.TCGA.SARC,expression.TCGA.SARC,expression.GSE71119)

gc()

```

Now that we have the adequate coding of the clinical data, we proceed to make testing and training sets from both cohorts, taking into consideration the disparity of events in both datasets. 

```{r create training and testing sets}

table(clinical.cinsarc$event)

table(clinical.tcga$event)

cinsarc <- generate.testing.training.sets(list(clinical=clinical.cinsarc,expression=for.modeling.expression.cinsarc))

tcga <- generate.testing.training.sets(list(clinical=clinical.tcga,expression=for.modeling.expression.tcga))

tcga.training <- cbind(tcga$training[["clinical"]], tcga$training[["expression"]])

tcga.testing <- cbind(tcga$testing[["clinical"]], tcga$testing[["expression"]])

cinsarc.training <- cbind(cinsarc$training[["clinical"]], cinsarc$training[["expression"]])

cinsarc.testing <- cbind(cinsarc$testing[["clinical"]], cinsarc$testing[["expression"]])

```

```{r delete temporary variables 3, include=FALSE}

rm(for.modeling.expression.tcga,for.modeling.expression.cinsarc,clinical.tcga,clinical.cinsarc, cinsarc, tcga)

gc()

```

Now, finally we are ready to do run some analytics on the datasets. We will run individual random forests in the two training datasets and then boost them together to hopefully have greater overall accuracy.

```{r running random forest and making the ensemble}

rF.tcga <- run.classifier (tcga.training, classifier = "random forest")

varImpPlot(rF.tcga)

evaluate.classifier (tcga.testing, rF.tcga)

rF.cinsarc <- run.classifier (cinsarc.training, classifier = "random forest")

varImpPlot(rF.cinsarc)

evaluate.classifier (cinsarc.testing ,rF.cinsarc)

bagging.rFs(cinsarc.training, tcga.training)
```

We have demonstrated the use of Random Forests for modeling the genetic determinants for sarcoma disease development. Up to this point, and as demonstrated by the ROC curves, this model is still not good enough to determine which are the genes that would be the best candidates for the development of a GEP product. We have still other alternatives, such as the use of SVM and other type of classifiers, as well as integrating other patient cohorts as they become available.

```{r run svms, eval=FALSE, include=FALSE}

#svm.tcga <- run.classifier (cbind(tcga$training[["clinical"]],tcga$training[["expression"]]), classifier = "svm")

#predict.svm.tcga <-  predict(svm.tcga, cbind(tcga$testing[["clinical"]], tcga$testing[["expression"]]))

#svm.cinsarc <- run.classifier (cbind(cinsarc$training[["clinical"]], cinsarc$training[["expression"]]), classifier = "svm")

#predict.svm.cinsarc <-  predict(svm.cinsarc, cbind(cinsarc$testing[["clinical"]], cinsarc$testing[["expression"]]))

```

```{r ensemble svms and random forests, eval=FALSE, include=FALSE}

#predict.cinsarc <- ensembl.svm.rF(predict.rF.cinsarc ,predict.svm.cinsarc) 

#predict.tcga <- ensembl.svm.rF(predict.rF.tcga, predict.svm.tcga) 

```
