---
title: "Springboard Foundations of Data Science"
author: "Rocio Dominguez Vidana, PhD; and Matt Fornito, PhD (mentor)"
date: "August 30, 2016"
output: html_document
---

```{r setup, include=FALSE}

knitr::opts_chunk$set(echo = TRUE)

```

# Development of Gene Expression Profile signature to predict survival in Sarcoma patients (Capstone Project)

Molecular diagnostics is the area at the center of a healthcare revolution and under intense development. It is possible by the increasing use  of powerful profiling tools that leverage our understanding of genetic variability within individuals in healthy and diseased states, and it can inform treatment and prognosis for patients.

The client company has extensive experience developing prognostic tests with “gene expression signatures” in a subset of rare neoplastic diseases for which there are not a lot of research resources allocated to. Using a combination of publically available data curated with further validation. They have developed products which can be used to predict metastatic potential, and thus inform aggressivity of treatment. The company's objective is to achieve that these become “standard of care” by improving the patient's diagnostic; and thus, be rutinarily reimbursed by insurance companies.

They have decided to focus on developing a product for sarcomas, a rare type of cancer in humans that arises from transformed cells of mesenchymal origin (bone, cartilage, fat, muscle, vascular, hematopoietic tissues, etc). Within sarcomas there is extreme genetic variability, 30% of them have distinct genetic features that can be considered the cause of those particular tumors. However, for the other 70% percent, there is really no traceable genetic modification. Currently, there are two publically available datasets:

RNASeq TCGA-SARC dataset from [GDC Data Portal](https://gdc-portal.nci.nih.gov/) data (in log2(normalized_count+1) units).

This will be called TCGA dataset

[Expression data from Complex genetics sarcomas (cohort 1 and 2)](http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE21050)  The data was measured in Affymetrix Human Genome U133 Plus 2.0 Array

This will be called CINSARC dataset
however there are many challenges in them like 


Load the two available RNA expression datasets, and the function script.

```{r load dataset}

load("capstone.RData")

source("capstone.code.R")

dim(expression.GSE71119)

dim(expression.TCGA.SARC)

```

There are different subsets of genes available in each dataset, in order to determine the best predictors accross both datasets, we will keep only the common genes. 

```{r identify common genes}

common.expression.genes <- intersect(colnames(expression.TCGA.SARC),colnames(expression.GSE71119))

```

A transformation to a normal distribution by removing the average effect of both all the genes, and all the samples. 
```{r normalize datasets}

normalized.tcga <- normalize.expression.data(expression.TCGA.SARC[,common.expression.genes])

normalized.cinsarc <- normalize.expression.data(expression.GSE71119[,common.expression.genes])

```

Then we will take the top 2/3 of the genes with the highest standard deviation, and we remove the less expressed genes (only the bottom 10% that has a gene mean expression z score lesser than  .5.

```{r select common normalized genes}

selected.tcga <- select.variable.genes (normalized.tcga)

selected.cinsarc <- select.variable.genes (normalized.cinsarc)

common.variable.genes<-intersect(selected.tcga$genes,selected.cinsarc$genes)

```

Then we remove all the genes with correlation higher than .5

```{r choose workable genes, reduce dimensionality}

valid.genes.tcga <- remove.correlated.genes.1(list(genes.rank = selected.tcga$genes.rank, expression.data = normalized.tcga[,common.variable.genes]))

valid.genes.cinsarc <- remove.correlated.genes.1(list(genes.rank = selected.cinsarc$genes.rank, expression.data = normalized.cinsarc[,common.variable.genes]))

common.valid.genes<-intersect(valid.genes.tcga,valid.genes.cinsarc)

for.modeling.expression.cinsarc <- normalized.cinsarc[,common.valid.genes]

for.modeling.expression.tcga <- normalized.tcga[,common.valid.genes]

```

```{r delete bunch of intermediate variables, include=FALSE}

rm(common.expression.genes,common.valid.genes,common.variable.genes,selected.cinsarc,selected.tcga,valid.genes.tcga,valid.genes.cinsarc)

gc()

```

Now we are focusing on the clinical data this 

```{r data janitoring part 1}

clinical.cinsarc <- clinical.GSE71119[,c(8:10)]
colnames(clinical.cinsarc) <- c("subtype","event","time")
clinical.cinsarc$subtype <- as.character(clinical.cinsarc$subtype)
clinical.cinsarc$event <- as.character(clinical.cinsarc$event)
clinical.cinsarc$time <- as.numeric(as.character(clinical.cinsarc$time))*30

clinical.tcga <- clinical.TCGA.SARC[,c(2,9,32)]
colnames(clinical.tcga) <- c("event","time","subtype")
clinical.tcga$time <- as.numeric(clinical.tcga$time)
clinical.tcga$subtype <- as.character(clinical.tcga$subtype)

```

```{r data janitoring part 2, include=FALSE}
clinical.cinsarc$subtype[which(clinical.cinsarc$subtype=="Dedifferentiated liposarcoma")]<-rep("1",length(which(clinical.cinsarc$subtype=="Dedifferentiated liposarcoma")))
clinical.cinsarc$subtype[which(clinical.cinsarc$subtype=="Leiomyosarcoma")]<-rep("2",length(which(clinical.cinsarc$subtype=="Leiomyosarcoma")))
clinical.cinsarc$subtype[which(clinical.cinsarc$subtype=="GIST")]<-rep("2",length(which(clinical.cinsarc$subtype=="GIST")))
clinical.cinsarc$subtype[which(clinical.cinsarc$subtype=="Myxofibrosarcoma")]<-rep("3",length(which(clinical.cinsarc$subtype=="Myxofibrosarcoma")))
clinical.cinsarc$subtype[which(clinical.cinsarc$subtype=="Undifferentiated sarcoma")]<-rep("4",length(which(clinical.cinsarc$subtype=="Undifferentiated sarcoma")))
clinical.cinsarc$subtype[which(clinical.cinsarc$subtype=="Other")]<-rep("6",length(which(clinical.cinsarc$subtype=="Other")))
clinical.cinsarc$subtype <- as.integer(clinical.cinsarc$subtype)
clinical.cinsarc$event[which(clinical.cinsarc$event=="Yes")]<-rep("1",length(which(clinical.cinsarc$event=="Yes")))
clinical.cinsarc$event[which(clinical.cinsarc$event=="No")]<-rep("0",length(which(clinical.cinsarc$event=="No")))
clinical.cinsarc$event <- as.integer(clinical.cinsarc$event)

clinical.tcga$subtype[which(clinical.tcga$subtype=="Dedifferentiated liposarcoma")]<-rep("1",length(which(clinical.tcga$subtype=="Dedifferentiated liposarcoma")))
clinical.tcga$subtype[which(clinical.tcga$subtype=="Leiomyosarcoma (LMS)")]<-rep("2",length(which(clinical.tcga$subtype=="Leiomyosarcoma (LMS)")))
clinical.tcga$subtype[which(clinical.tcga$subtype=="Myxofibrosarcoma")]<-rep("3",length(which(clinical.tcga$subtype=="Myxofibrosarcoma")))
clinical.tcga$subtype[which(clinical.tcga$subtype=="Pleomorphic MFH/ Undifferentiated pleomorphic sarcoma")]<-rep("4",length(which(clinical.tcga$subtype=="Pleomorphic MFH/ Undifferentiated pleomorphic sarcoma")))
clinical.tcga$subtype[which(clinical.tcga$subtype=="Giant cell MFH / Undifferentiated pleomorphic sarcoma with giant cells")]<-rep("4",length(which(clinical.tcga$subtype=="Giant cell MFH / Undifferentiated pleomorphic sarcoma with giant cells")))
clinical.tcga$subtype[which(clinical.tcga$subtype=="Undifferentiated Pleomorphic Sarcoma (UPS)")]<-rep("4",length(which(clinical.tcga$subtype=="Undifferentiated Pleomorphic Sarcoma (UPS)")))
clinical.tcga$subtype[which(clinical.tcga$subtype=="Sarcoma; synovial; poorly differentiated")]<-rep("5",length(which(clinical.tcga$subtype=="Sarcoma; synovial; poorly differentiated")))
clinical.tcga$subtype[which(clinical.tcga$subtype=="Synovial Sarcoma - Biphasic")]<-rep("5",length(which(clinical.tcga$subtype=="Synovial Sarcoma - Biphasic")))
clinical.tcga$subtype[which(clinical.tcga$subtype=="Synovial Sarcoma - Monophasic")]<-rep("5",length(which(clinical.tcga$subtype=="Synovial Sarcoma - Monophasic")))
clinical.tcga$subtype[which(clinical.tcga$subtype=="Desmoid Tumor")]<-rep("6",length(which(clinical.tcga$subtype=="Desmoid Tumor")))
clinical.tcga$subtype[which(clinical.tcga$subtype=="Malignant Peripheral Nerve Sheath Tumors (MPNST)")]<-rep("6",length(which(clinical.tcga$subtype=="Malignant Peripheral Nerve Sheath Tumors (MPNST)")))

clinical.tcga$subtype <- as.integer(clinical.tcga$subtype)

rm(clinical.GSE71119,clinical.TCGA.SARC,expression.TCGA.SARC,expression.GSE71119)
save.image("C:/Users/rvidana/Dropbox/capstone/capstone.work.RData")
gc()

```



```{r create training and testing sets}
cinsarc <- generate.testing.training.sets(list(clinical=clinical.cinsarc,expression=for.modeling.expression.cinsarc))
tcga <- generate.testing.training.sets(list(clinical=clinical.tcga,expression=for.modeling.expression.tcga))
rm(for.modeling.expression.tcga,for.modeling.expression.cinsarc,clinical.tcga,clinical.cinsarc)

```



```{r running random forest}

#glm.tcga <- glm(event ~ . - time ,data=cbind(tcga$training[["clinical"]],tcga$training[["expression"]]),family=binomial())

#glm.cinsarc <- glm(event ~ . - time ,data=cbind(cinsarc$training[["clinical"]],cinsarc$training[["expression"]]),family=binomial())

rF.tcga <- randomForest(event ~ . - time, data = cbind(tcga$training[["clinical"]],tcga$training[["expression"]]))

predict.rF.tcga <-  predict(rF.tcga, cbind(tcga$testing[["clinical"]],tcga$testing[["expression"]]))

rpart.tcga <- rpart(event ~ . - time, data = cbind(tcga$training[["clinical"]],tcga$training[["expression"]]),method="class")

predict.rpart.tcga <-  predict(rpart.tcga, cbind(tcga$testing[["clinical"]],tcga$testing[["expression"]]))

rF.cinsarc <- randomForest(event ~ . - time, data = cbind(cinsarc$training[["clinical"]],cinsarc$training[["expression"]]))

predict.rF.cinsarc <-  predict(rF.cinsarc, cbind(cinsarc$testing[["clinical"]],cinsarc$testing[["expression"]]))

rpart.cinsarc <- rpart(event ~ . - time, data = cbind(cinsarc$training[["clinical"]],cinsarc$training[["expression"]]),method="class")

predict.rpart.cinsarc <-  predict(rpart.cinsarc, cbind(cinsarc$testing[["clinical"]],cinsarc$testing[["expression"]]))

varImpPlot(rF.tcga)

```

run svms

```{r run svms}
svm.tcga <- svm(event ~ . - time, data = cbind(tcga$training[["clinical"]],tcga$training[["expression"]]))

predict.svm.tcga <-  predict(svm.tcga, cbind(tcga$testing[["clinical"]],tcga$testing[["expression"]]))

svm.cinsarc <- svm(event ~ . - time, data = cbind(cinsarc$training[["clinical"]],cinsarc$training[["expression"]]))

predict.svm.cinsarc <-  predict(svm.cinsarc, cbind(cinsarc$testing[["clinical"]],cinsarc$testing[["expression"]]))
```
