---
title: "Springboard Foundations of Data Science"
author: "Rocio Dominguez Vidana, PhD"
mentor: "Matt Fornito, PhD"
date: "July 13, 2016"
output: pdf_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Capstone Project

##  Gene expression Profiling analysis

Molecular diagnostic is nowadays possible using powerful profiling tools that can inform treatment and prognosis for patients. however there are many challenges in them like 


Load the two available RNA expression datasets.

```{r}
load("capstone.RData")
dim(rnaseq.GSE71119)
dim(hiseq.TCGA.SARC)
```

There are different subsets of genes available in each dataset, since we intend on using one dataset to ascertain the robustnez of of the findings of the other dataset, we will keep only the common genes. 

```{r}
rnaseq.cinsarc.common<-rnaseq.GSE71119[,which(colnames(rnaseq.GSE71119)%in%colnames(rnaseq.TCGA.SARC))]
rnaseq.common<-rnaseq.TCGA.SARC[,which(colnames(rnaseq.TCGA.SARC)%in%colnames(rnaseq.GSE71119))]
dim(rnaseq.common)
dim(hiseq.common)
```

We will focus on the first dataset, and transform it into a normal distribution. Then we will take the top 2/3 of the data with the highest standard deviation.

```{r}
hist(as.matrix(rnaseq.common)) 
tmp<-rnaseq.common
tmp[is.na(tmp)] <- 0
tmp.2<-tmp-apply(tmp,2,median)
rnaseq.norm<-tmp.2-apply(tmp.2,1,median)
rm(tmp)
rm(tmp.2)
hist(as.matrix(rnaseq.norm))

rnaseq.sd<-apply(rnaseq.norm,2,sd)
hist(rnaseq.sd)
quantile(rnaseq.sd)
rnaseq.sd.ecdf<-ecdf(rnaseq.sd)
rnaseq.sd.ecdf(4.74) #cinscarc
rnaseq.sd.ecdf(4.03) #tcga
#remove the 1/3 with less standard deviation #4.68 for cinsarc
#maybe change to 1/4
length(which(rnaseq.sd>=4.68))

rnaseq.mean<-apply(rnaseq.norm,2,mean)
rnaseq.zmean<-(rnaseq.mean-mean(rnaseq.mean))/sd(rnaseq.mean)
hist(rnaseq.zmean)
plot(rnaseq.mean,rnaseq.zmean)
plot(rnaseq.sd,rnaseq.zmean)
#remove the probes with very little expression
length(which(abs(rnaseq.zmean)>=2)) 

good.genes<-which(abs(rnaseq.zmean)>=1 & rnaseq.sd>=4.68) #cinscarc
ok.genes<-which(abs(rnaseq.zmean)>=.5 & rnaseq.sd>=4) #cinscarc roughy bottom 10% with less sd, 
ok.genes<-which(abs(rnaseq.zmean)>=.5 & rnaseq.sd>=3.68) #cinscarc roughy bottom 10% with less sd, 

#rank probes absolutes and in descending order, remove the genes with which that have absolute .75 correlation with other genes
ok.genes.rank<-rank(abs(rnaseq.zmean[ok.genes]))
tmp<-cor(rnaseq.norm[,ok.genes])
tmp[lower.tri(tmp,diag=TRUE)]<-NA
tmp[abs(tmp) <= 0.5] <- NA
tmp.2<-as.data.frame(as.table(tmp))
rm(tmp)
colnames(tmp.2)<-c('gene_1','gene_2','value')
tmp<-na.omit(tmp.2)
rm(tmp.2)
tmp<-tmp[order(-abs(tmp$value)),]
tmp$rank_1<-ok.genes.rank[tmp$gene_1]
tmp$rank_2<-ok.genes.rank[tmp$gene_2]
tmp$remove<-rep('NA',dim(tmp)[1])
tmp$remove<-ifelse((tmp$rank_1 >= tmp$rank_2),as.character(tmp$gene_1),as.character(tmp$gene_2))
keep.genes<-which(!(names(ok.genes)%in%remove.genes))
rm(tmp)

#do it for dataset 2
tmp<-.common
tmp[is.na(tmp)] <- 0
tmp.2<-tmp-apply(tmp,2,median)
rnaseq.norm<-tmp.2-apply(tmp.2,1,median)


```


```{r}
#
```
